### PR TITLE
feat(auth): Add multi-factor support for the sign-in flow

### PR DESCRIPTION
**Pull request by fzuellich**
### Description

At Kyan Health we are using `react-native-firebase` for our mobile apps. We plan to roll out multi-factor authentication to our users soon. After implementing the necessary code on the web (using the original Firebase libraries) we were missing multi-factor support for the apps.

I'm aware that there are a few small todos left in the code. Nonetheless I'd appreciate some early feedback from the maintainers to see how we can get the PR merged. :)

#### Open questions on our side:
* I'm not sure if I'm supposed to add a license header to new files. Let me know and I'm happy to copy the existing header into the newly created files.
*  I'm running into a lot of trouble with testing the iOS part against the Firebase emulator. I marked these tests with `:android:`. I have the feeling that this is not properly implemented on Firebase' side. Testing against a real life Firebase project doesn't run into issues with missing multi-factor authentication. I also find it strange the the Firebase emulator prints a different SMS code message for iOS devices (masks the phone number). Any thought or hints how to solve this problem? 😇 

### Related issues

I'm aware of:

* #4166 Adding multi-factor information to the user object IMHO makes only sense if sign-in is supported, otherwise I believe it is not possible to obtain a valid session and the multi-factor information.
* #5372 Does provide multi-factor enrolment and sign-in flow. I believe our PR follows the official API closer (https://cloud.google.com/identity-platform/docs/web/mfa), but implements only sign-in and skips the some of the error message conversion.

### Checklist

- I read the [Contributor Guide](../CONTRIBUTING.md) and followed the process outlined there for submitting PRs.
  - [x] Yes
- My change supports the following platforms;
  - [x] `Android`
  - [x] `iOS`
- My change includes tests;
  - [x] `e2e` tests added or updated in `packages/\*\*/e2e`
  - [x] `jest` tests added or updated in `packages/\*\*/__tests__`
- [x] I have updated TypeScript types that are affected by my change.
- This is a breaking change;
  - [ ] Yes
  - [x] No

### Test Plan

* e2e tests run
* I could verify the multi-factor enrollment and sign-in with a real world Firebase project (iOS)
🔥

### CODE REVIEW

File: docs/auth/phone-auth.md
Comment (fzuellich) [inline docs/auth/phone-auth.md:4]: Typo, this should be
```suggestion
next: /auth/mfa-auth
```

File: packages/auth/lib/index.js
Comment (fzuellich) [inline packages/auth/lib/index.js:32]: Slipped in, needs to go

File: packages/auth/lib/index.js
Comment (fzuellich) [inline packages/auth/lib/index.js:252]: Slipped in, needs to go

File: docs/auth/mfa-auth.md
Comment (fzuellich) [inline docs/auth/mfa-auth.md:2]: This file does not show up in TOC yet

**Comment by mikehardy** [comment]
I'm really sorry I haven't gotten to this yet - I *really* want this functionality, I truly appreciate the PR, and I will collaborate with you to get it merged+released. I appreciate the patience

File: packages/auth/ios/RNFBAuth/RNFBAuthModule.m
Comment (fzuellich) [inline packages/auth/ios/RNFBAuth/RNFBAuthModule.m:1435]: Bug: This returns an array of `null`.

File: packages/auth/e2e/multiFactor.e2e.js
Comment (mikehardy) [inline packages/auth/e2e/multiFactor.e2e.js:294]: firebase-tools repo (which hosts the emulator) appears to be tracking this - I just chimed in over there: https://github.com/firebase/firebase-tools/issues/4290#issuecomment-1281260335 - could probably link it here

```suggestion
      // This test should probably be implemented but doesn't work:
      // Every time the same phone number requests a verification code,
      // the emulator endpoint does not return a code, even though the emulator log
      // prints a code.
      // See https://github.com/firebase/firebase-tools/issues/4290#issuecomment-1281260335
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
Wow, this looks like excellent work. Finally got a chance to review through it.
I left some mostly-trivial / easy to resolve comments, what do you think?

How is this working for you currently? Does it seem to work? You may not have had a chance to test integrate it yet as the CI runs require approval for first-time PRs so the patch-package patch set is not being generated until I approve.

I just approved a run now which should kick out patches for test integration case that helps

**Comment by mikehardy** [comment]
Oh - should have written this as nearly the first thing: the test support is absolutely fantastic, really really appreciated. It gives me so much more confidence during review that it does in fact work, knowing it's got e2e tests going with coverage we can see

**Comment by mikehardy** [comment]
Interesting jest failure, referencing appCheck package? Is that a transitive dep of MFA? :thinking: 

```

  ● Auth › getMultiFactorResolver › should return the resolver object if its found

    You attempted to use "firebase.app('[DEFAULT]').appCheck" but this module could not be found.

    Ensure you have installed and imported the '@react-native-firebase/app-check' package.
```

Formatting failure can be fixed with 

https://github.com/invertase/react-native-firebase/blob/bf1de6c5863e065df0cddd1125b61e1bafebb927/package.json#L13

(likewise if there is an android one, just run `yarn lint:android` locally - possibly a few times until it passes, then commit the result, and for javascript `yarn lint:js --fix` and/or `yarn lint:markdown --write` I think - I forget exactly, but basically getting those to not just check but also fix things in place, then commit result

File: packages/auth/e2e/multiFactor.e2e.js
Comment (fzuellich) [reply packages/auth/e2e/multiFactor.e2e.js]: I took a look at the firebase-js-sdk, there they fixed it at one point: https://github.com/firebase/firebase-js-sdk/commit/645e1668ed5783fce9a3cfe7efe1c9a044280046

I can't find the error message in the Android SDK. To be honest, I'm not exactly sure where the error message comes from. Straight from the API/Emulator? I changed the assertion 👍

File: packages/auth/e2e/multiFactor.e2e.js
Comment (fzuellich) [reply packages/auth/e2e/multiFactor.e2e.js:196]: Yes this is one of the TODOs I mentioned. There is at least one more in the Android code.

File: packages/auth/e2e/multiFactor.e2e.js
Comment (mikehardy) [reply packages/auth/e2e/multiFactor.e2e.js:62]: Let me see what the output looks like after the e2e runs, as long as there is still evidence in the output that there is a test but it is skipped (and the total test count / skipped test count registers them), that's the real goal for me. How they are skipped isn't so important

Comment (mikehardy) [reply packages/auth/e2e/multiFactor.e2e.js:62]: It does not appear to show up in list of skipped tests (it would be an entry with a `-`) so I don't think this works well - it hides them too well. I'd prefer the if-platform/else-skip style

```
  multi-factor
    ✔ has no multi-factor information if not enrolled
    ✔ can not reuse session after logout (1035ms)
    enroll
      ✔ throws an error for unknown sessions
      ✔ throws an error for unknown verification code
```
https://github.com/invertase/react-native-firebase/actions/runs/3271818247/jobs/5385708987#step:23:282

**Comment by mikehardy** [comment]
You may want to rebase this against current main - I've just ingested the major releases firebase-ios-sdk v10 and firebase-android-sdk v31, and on android they did fix a multi-factor issue:

https://firebase.google.com/support/release-notes/android#auth_v21-1-0

The releases themselves are otherwise very low impact (in my opinion), with the only actual breaking change that affects react-native-firebase being a higher minimum iOS target (11) now

I enhanced the android e2e stability as well so CI shouldn't be so finicky.

Other than that, I'll stand by until I hear the "all clear" then I can test through this again. No pressure at all of course, but I am excited to see it merged, whenever it is ready. Cheers

**Comment by fzuellich** [comment]
> You may want to rebase this against current main - I've just ingested the major releases firebase-ios-sdk v10 and firebase-android-sdk v31, and on android they did fix a multi-factor issue:
> 
> https://firebase.google.com/support/release-notes/android#auth_v21-1-0

Awesome, that explains the broken timestamp when using Android! Do you want me to do a rebase and push-force or rather merge main into this branch?

We have some urgent stuff going right now, so I get less done than I would like.

**Comment by mikehardy** [comment]
I have to admit I'm a force-push (or, more specifically `git push --force-with-lease`) sort of person that prefers clean linear commit histories. But I don't force my opinions on people everywhere, especially since it implies a level of git that is past what is strictly required to produce still-useful code

So pull rebase + force-push it back if you can and if you want, but merges can work as well

**Comment by fzuellich** [comment]
Happy to do the rebase.  I'm asking, because the only problem I have with push-force to Github PRs is, that Github doesn't understand it so well and review comments might get lost. :)

I'm a fan of a clean history as well. Do you plan on squashing and merging this PR? I feel like I added some "not so useful" commits here. Otherwise I'd squash the less useful commits during the rebase and then push force.

**Comment by mikehardy** [comment]
I'm used to the commit disassociation thing yes - no big deal there
I choose squash or rebase on a PR-by-PR basis depending on how the history looks - if you *want* to make a clean history (and your git-fu is strong enough it's no trouble) by all means go ahead and rebase locally to clean it up then I will rebase, but I can squash too.

I personally make clean linear commit PRs with independently testable / revertable commits but that's more of a compulsion than a requirement certainly, it's not a blocker here

In the end it's the green CI + green review on the last commit and the rest is details I won't get stuck on, "implementor's choice"

File: packages/auth/e2e/multiFactor.e2e.js
Comment (fzuellich) [reply packages/auth/e2e/multiFactor.e2e.js]: Done :)

**Comment by fzuellich** [comment]
I rebased on invertase/main and "fixed" the [appCheck problem](https://github.com/invertase/react-native-firebase/pull/6593#issuecomment-1281278777). I'm don't understand enough of JavaScript to be able to explain how `expect.toEqual` and react-native-firebase interact to produce that error. But comparing the properties works fine on my machine.

Currently this test https://github.com/invertase/react-native-firebase/pull/6593/files#diff-50df2eada48dcacc67e8af24c907da9a349aa7fb0fd2e3ed01aaf0258a2f04baR30 fails, because I spotted a bug in the test implementation. @mikehardy Do you think this is a useful test? I'm not sure. If an attacker knows the session and has access to the users phone to receive the second factor, then the attacker can gain access to a valid session without knowing the password.

I'll keep working on the todos in the code. Not quite ready yet.

**Comment by mikehardy** [comment]
Test usefulness depends on the perspective taken - you are probably correct with regards to security surface area though I'm always cautious with security things. From a maintenance perspective it's best to actually test every bit of the contract and if session is supposed to be invalid after logout, it's good to probe that. I'm always amazed by what unrelated things I find when I really probe all parts of an API contract! That thought combined with the idea that security surface area may be incorrectly analyzed (for whatever reason) makes me think it is useful. I would hate to be the attack vector that caused a real world life+safety issue for someone because of a security issue :see_no_evil:

**Comment by fzuellich** [comment]
The latest commits should fix all todos for this PR. I double checked linting 🤞 

Currently I'm unable to integrate the latest changes into our app using yalc. Some classpath issues for Android and iOS. I hope this is resolved when I switch to a patch-package. 

Please test through this again. :)

**Comment by mikehardy** [comment]
Huh - never heard of yalc before, interesting! Either way, CI running now, should be a new patch-package set shortly and since I just released (again) last night the patch set will be clean

**Comment by fzuellich** [comment]
I did some more testing and discovered two smaller issues with the error handling. I pushed the fixes as two commits. I hope it makes it easier to review the changes.

**Comment by mikehardy** [comment]
:thinking: 

```
❌ /Users/runner/work/react-native-firebase/react-native-firebase/packages/auth/ios/RNFBAuth/RNFBAuthModule.m:786:1: extraneous closing brace ('}')
}
```

File: packages/auth/e2e/multiFactor.e2e.js
Comment (fzuellich) [reply packages/auth/e2e/multiFactor.e2e.js]: The todos are sovled now :)

**Comment by fzuellich** [comment]
Hola @mikehardy, I pushed a fix for that syntax error. If you get a chance can you let the checks run again? :)

**Comment by mikehardy** [comment]
Green CI is a beautiful thing - I'll do one more review pass through here but I imagine this is good to go if it is working for you - the patch-set may be useful in verifying same ?

**Comment by fzuellich** [comment]
I'm quite happy with the iOS part so far (using the patch set). Still having trouble with the Android side due to 

`Duplicate class com.google.firebase.iid.FirebaseInstanceIdReceiver found in modules jetified-firebase-iid-20.0.1-runtime (com.google.firebase:firebase-iid:20.0.1) and jetified-firebase-messaging-23.1.0-runtime (com.google.firebase:firebase-messaging:23.1.0)`

So far the internet wasn't too useful to fix this issue and I secretly hope this will magically solve itself once we have a real package. 🙈

**Comment by mikehardy** [comment]
That one is a pain!

https://github.com/invertase/react-native-firebase/blob/e6fbf592e1014508528b302e4aa4cc63f052ddd1/packages/messaging/android/build.gradle#L119-L120

https://github.com/firebase/firebase-android-sdk/issues/4206

**Comment by fzuellich** [comment]
Thanks Mike, your suggestion fixed it! I need to improve my Google powers ;)

I walked through the enrollment and sign-in using Android as well. It works for us 🥳

**Comment by mikehardy** [comment]
Fantastic, so my status understanding is no known issues/ read for final review+merge release (hopefully) then. Enqueued for me as such

**Comment by fzuellich** [comment]
Sounds good. I'm sure we'll be the first to open a PR if we find a bug. 😇  Do you have an ETA that I can communicate to management?

**Review by mikehardy — APPROVED** [review]
Re-scanned it all for final review ✅ 
Tested e2e locally after pulling current main + rebase (just to be sure)
- ios ✅ 
- android ✅ 

Merging now, will release in just a moment so should be available as 16.3.0 in a few minutes.

You can communicate to management that the quality of the PR and attention to detail on testing gives me a really high level of confidence, making quick review possible and leading to me shipping it immediately despite the size of the PR, how's that vs an ETA ;-)?. This is a fantastic addition to the module, I really appreciate it and I'm sure others hoping for it will be thrilled as well. Happy to collaborate on any followups as needed if (or when, it's a big feature...) something comes up. Cheers

**Comment by fzuellich** [comment]
That sounds awesome. Thank you so much for the collaboration!

**Comment by mikehardy** [comment]
@fzuellich not sure if this is relevant for you or not, but saw it in the release notes for just-released 10.1.0 of firebase-ios-sdk and thought you might be interested, it just released: https://github.com/firebase/firebase-ios-sdk/pull/10296

I'm releasing the SDK update packaged into a version here shortly so that fix (among others) will be generally available here

**Pull request merged**
